BaragonService/src/main/java/com/hubspot/baragon/service/elb/ApplicationLoadBalancer.java

```diff
@@ -0,0 +1,542 @@
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Sets;
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
+import com.hubspot.baragon.data.BaragonKnownAgentsDatastore;
+import com.hubspot.baragon.data.BaragonLoadBalancerDatastore;
+import com.hubspot.baragon.models.BaragonAgentMetadata;
+import com.hubspot.baragon.models.BaragonGroup;
+import com.hubspot.baragon.models.TrafficSource;
+import com.hubspot.baragon.models.TrafficSourceType;
+import com.hubspot.baragon.service.BaragonServiceModule;
+import com.hubspot.baragon.service.config.ElbConfiguration;
+import com.hubspot.baragon.service.exceptions.BaragonExceptionNotifier;
+
+/**
+ * Handles interactions with the ApplicationLoadBalancer in AWS.
+ *
+ * For this class I am working under the assumption that the ID's of Targets can
+ * be interchanged with the ID's of instances. That is, if you ask this class
+ * for the health of instance 12345, then you will actually expect the health
+ * of target 12345.
+ */
+public class ApplicationLoadBalancer extends ElasticLoadBalancer {
+  private static final Logger LOG = LoggerFactory.getLogger(ApplicationLoadBalancer.class);
+  private final AmazonElasticLoadBalancingClient elbClient;
+
+  @Inject
+  public ApplicationLoadBalancer(Optional<ElbConfiguration> configuration,
+                                 BaragonExceptionNotifier exceptionNotifier,
+                                 BaragonLoadBalancerDatastore loadBalancerDatastore,
+                                 BaragonKnownAgentsDatastore knownAgentsDatastore,
+                                 @Named(BaragonServiceModule.BARAGON_AWS_ELB_CLIENT_V2) AmazonElasticLoadBalancingClient elbClient) {
+    super(configuration, exceptionNotifier, loadBalancerDatastore, knownAgentsDatastore);
+    this.elbClient = elbClient;
+  }
+
+  @Override
+  public boolean isInstanceHealthy(String instanceId, String elbName) {
+    DescribeTargetHealthRequest healthRequest = new DescribeTargetHealthRequest()
+        .withTargets(new TargetDescription().withId(instanceId));
+    DescribeTargetHealthResult result = elbClient.describeTargetHealth(healthRequest);
+
+    for (TargetHealthDescription health: result.getTargetHealthDescriptions()) {
+      if (health.getTargetHealth().getState().equals(TargetHealthStateEnum.Healthy.toString())
+          && health.getTarget().getId().equals(instanceId)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  @Override
+  public void removeInstance(Instance instance, String trafficSourceName, String agentId) {
+    Optional<TargetGroup> maybeTargetGroup = getTargetGroup(trafficSourceName);
+    if (maybeTargetGroup.isPresent()) {
+      TargetGroup targetGroup = maybeTargetGroup.get();
+      TargetDescription targetDescription = new TargetDescription()
+          .withId(instance.getInstanceId());
+
+      DeregisterTargetsRequest deregisterTargetsRequest = new DeregisterTargetsRequest()
```

Comment:
Ah that's right, I had it in my head that the list already came with the target group. 👍